Signal/src/util/OWSBackup.m

```diff
@@ -0,0 +1,908 @@
+            [dateFormatter setDateFormat:@".yyyy.MM.dd hh.mm.ss"];
+            NSString *replacementDateTime = [dateFormatter stringFromDate:[NSDate new]];
+
+            NSString *oldFilePath = [dstFilePath stringByAppendingString:replacementDateTime];
+            BOOL success = [[NSFileManager defaultManager] moveItemAtPath:dstFilePath toPath:oldFilePath error:&error];
+            if (!success || error) {
+                OWSFail(@"%@ failed to move directory item: %@, %@", self.logTag, dstFilePath, error);
+                return NO;
+            }
+            if (![OWSFileSystem protectFolderAtPath:oldFilePath]) {
+                OWSFail(@"%@ failed to protect old directory item: %@, %@", self.logTag, oldFilePath, error);
+                return NO;
+            }
+        }
+
+        BOOL success = [[NSFileManager defaultManager] moveItemAtPath:srcFilePath toPath:dstFilePath error:&error];
+        if (!success || error) {
+            OWSFail(@"%@ failed to move directory item: %@, %@", self.logTag, dstFilePath, error);
+            return NO;
+        }
+        if (![OWSFileSystem protectFolderAtPath:dstFilePath]) {
+            OWSFail(@"%@ failed to protect directory item: %@, %@", self.logTag, dstFilePath, error);
+            return NO;
+        }
+    }
+
+    return YES;
+}
+
+#pragma mark - Clean up
+
++ (void)cleanupBackupState
+{
+    DDLogInfo(@"%@ %s.", self.logTag, __PRETTY_FUNCTION__);
+
+    [self cleanupBackupDirectoriesInDirectory:NSTemporaryDirectory()];
+    [self cleanupBackupDirectoriesInDirectory:OWSFileSystem.appDocumentDirectoryPath];
+
+    [[NSUserDefaults appUserDefaults] removeObjectForKey:NSUserDefaults_QueuedBackupPath];
+    [[NSUserDefaults appUserDefaults] synchronize];
+}
+
++ (void)cleanupBackupDirectoriesInDirectory:(NSString *)dirPath
+{
+    OWSAssert(dirPath.length > 0);
+
+    NSError *error;
+    NSArray<NSString *> *filenames = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:dirPath error:&error];
+    if (error) {
+        OWSFail(@"%@ could not find files in directory: %@", self.logTag, error);
+        return;
+    }
+
+    for (NSString *filename in filenames) {
+        if (![filename hasPrefix:OWSBackup_DirNamePrefix]) {
+            continue;
+        }
+        NSString *filePath = [dirPath stringByAppendingPathComponent:filename];
+        BOOL success = [[NSFileManager defaultManager] removeItemAtPath:filePath error:&error];
+        if (!success || error) {
+            OWSFail(@"%@ could not clean up backup directory: %@", self.logTag, error);
+            return;
+        }
+    }
+}
+
+#pragma mark - Utils
+
+- (NSString *)relativePathforPath:(NSString *)filePath basePath:(NSString *)basePath
+{
+    OWSAssert(filePath.stringByStandardizingPath.length > 0);
+    OWSAssert([filePath.stringByStandardizingPath hasPrefix:basePath]);
+
+    NSString *relativePath = [filePath.stringByStandardizingPath substringFromIndex:basePath.length];
```

Comment:
Should we `basePath.stringByStandardizingPath.length` ?

Reply:
Done.